### PR TITLE
WS-027: Leaderboard Component

### PR DESCRIPTION
**Pull request by chiara-d697**
image

Leaderboard component created from LINQ join query and react table

### CODE REVIEW

File: backend/Repositories/UsersRepo.cs [hunk 1]
```diff
@@ -3,6 +3,7 @@
 using System.Linq;
 using WhaleSpotting.Models.Database;
 using WhaleSpotting.Models.Request;
+using  WhaleSpotting.Models.Response;
```
Comment (AndrewYHuang) [inline backend/Repositories/UsersRepo.cs:6]: Whitespace

File: backend/Repositories/UsersRepo.cs [hunk 2]
```diff
@@ -84,5 +87,30 @@ public User UpdateRole(int id, UpdateUserRoleRequest update)
...
+            foreach(var item in leaderboard)
+            {
+                leaderboardList.Add(new LeaderboardEntry{
+                    Username = item.username,
+                    Count = item.count});
+
+            }
+
+            return leaderboardList;
+
```
Comment (AndrewYHuang) [inline backend/Repositories/UsersRepo.cs:113]: Extra whitespace

File: backend/Repositories/UsersRepo.cs [hunk 3]
```diff
@@ -84,5 +87,30 @@ public User UpdateRole(int id, UpdateUserRoleRequest update)
...
+            _context.Sightings
+            .GroupBy(sighting => sighting.CreatedByUserId)
+            .Select(sighting => new {count = sighting.Count(), userId = sighting.Key})
+            .OrderByDescending(g => g.count)
+            .Join(_context.Users, sighting => sighting.userId, user => user.Id, (sighting, user) => 
+            new{username = user.Username, count = sighting.count});
+
+            List<LeaderboardEntry> leaderboardList = new List<LeaderboardEntry>();
+
+
```
Comment (AndrewYHuang) [inline backend/Repositories/UsersRepo.cs:103]: Extra whitespace

File: backend/Repositories/UsersRepo.cs [hunk 4]
```diff
@@ -84,5 +87,30 @@ public User UpdateRole(int id, UpdateUserRoleRequest update)
...
+        public List<LeaderboardEntry> GetLeaderboard()
+        {
+            var leaderboard = 
+            _context.Sightings
+            .GroupBy(sighting => sighting.CreatedByUserId)
+            .Select(sighting => new {count = sighting.Count(), userId = sighting.Key})
+            .OrderByDescending(g => g.count)
+            .Join(_context.Users, sighting => sighting.userId, user => user.Id, (sighting, user) => 
+            new{username = user.Username, count = sighting.count});
```
Comment (AndrewYHuang) [inline backend/Repositories/UsersRepo.cs:99]: I think we want some different indentation here

Comment (AndrewYHuang) [reply backend/Repositories/UsersRepo.cs:99]: Maybe something like
```suggestion
            return _context.Sightings
                .GroupBy(sighting => sighting.CreatedByUserId)
                .Select(sighting => new {count = sighting.Count(), userId = sighting.Key})
                .OrderByDescending(g => g.count)
                .Join(
                    _context.Users, 
                    sighting => sighting.userId, 
                    user => user.Id, 
                    (sighting, user) => new LeaderboardEntry { 
                        Username = user.Username, 
                        Count = sighting.count 
                    }
                ).ToList();
```

This way we can also remove quite a lot of the code below as we are returning a list of LeaderboardEntries

**Review by AndrewYHuang — CHANGES_REQUESTED** [review]

File: backend/Repositories/UsersRepo.cs (hunk 4 above)
Comment (chiara-d697) [reply backend/Repositories/UsersRepo.cs]: Refactored as suggested

File: backend/Repositories/UsersRepo.cs (hunk 1 above)
Comment (chiara-d697) [reply backend/Repositories/UsersRepo.cs]: Removed line

**Comment by chiara-d697** [comment]
Implemented the above fixes and resolved merge conflicts

**Review by TimLeach635 — APPROVED** [review]
Looks good to me!

**Pull request merged**